File: skills/plugin-creator/scripts/package_goose_plugin.py
#!/usr/bin/env python3
"""Validate and package a Goose/Open Plugins directory."""

import argparse
import subprocess
import sys
import time
import zipfile
from pathlib import Path

EXCLUDED_PARTS = {".git", ".hg", ".svn", "__pycache__", "node_modules"}
EXCLUDED_SUFFIXES = {".pyc", ".pyo"}
ZIP_EPOCH = 315532800


def should_exclude(path: Path, root: Path) -> bool:
    relative = path.relative_to(root)
    return any(part in EXCLUDED_PARTS for part in relative.parts) or path.suffix in EXCLUDED_SUFFIXES


def add_file(archive: zipfile.ZipFile, path: Path, archive_name: Path):
    if path.stat().st_mtime >= ZIP_EPOCH:
        archive.write(path, archive_name)
        return
    info = zipfile.ZipInfo(str(archive_name), time.localtime(ZIP_EPOCH)[:6])
    info.compress_type = zipfile.ZIP_DEFLATED
    info.external_attr = (path.stat().st_mode & 0xFFFF) << 16
    archive.writestr(info, path.read_bytes())


def main():
    parser = argparse.ArgumentParser(description="Validate and package a Goose/Open Plugins directory")
    parser.add_argument("plugin_dir")
    parser.add_argument("output", nargs="?", default=None)
    args = parser.parse_args()

    root = Path(args.plugin_dir).expanduser().resolve()
    validator = Path(__file__).with_name("validate_goose_plugin.py")
    subprocess.run([sys.executable, str(validator), str(root)], check=True)

    output = Path(args.output).expanduser().resolve() if args.output else root.parent / f"{root.name}.zip"
    output.parent.mkdir(parents=True, exist_ok=True)
    with zipfile.ZipFile(output, "w", compression=zipfile.ZIP_DEFLATED) as archive:
        for path in sorted(root.rglob("*")):
            if path.is_file() and not should_exclude(path, root) and path.resolve() != output:
                add_file(archive, path, Path(root.name) / path.relative_to(root))
    print(output)


if __name__ == "__main__":
    main()

File: skills/plugin-creator/scripts/package_goose_plugin.ts
#!/usr/bin/env node
// Validate and package a Goose/Open Plugins directory.
import { execFileSync } from "node:child_process";
import { mkdirSync, statSync, readdirSync } from "node:fs";
import { dirname, join, resolve, basename } from "node:path";
import { fileURLToPath } from "node:url";
import AdmZip from "adm-zip";

const HERE = dirname(fileURLToPath(import.meta.url));

const EXCLUDED_PARTS = new Set([".git", ".hg", ".svn", "__pycache__", "node_modules"]);
const EXCLUDED_SUFFIXES = new Set([".pyc", ".pyo"]);

function isDir(path: string): boolean {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}

function shouldExclude(path: string, root: string): boolean {
  const relative = path.slice(root.length + 1);
  const parts = relative.split(/[\\/]/);
  if (parts.some((p) => EXCLUDED_PARTS.has(p))) return true;
  const suffix = path.includes(".") ? path.slice(path.lastIndexOf(".")) : "";
  return EXCLUDED_SUFFIXES.has(suffix);
}

function collectFiles(root: string, current: string, out: string[]) {
  for (const entry of readdirSync(current).sort()) {
    const full = join(current, entry);
    if (isDir(full)) {
      collectFiles(root, full, out);
    } else {
      out.push(full);
    }
  }
}

function main() {
  const [pluginDirArg, outputArg] = process.argv.slice(2);
  if (!pluginDirArg) {
    console.error("usage: package_goose_plugin.js <plugin_dir> [output.zip]");
    process.exit(2);
  }

  const root = resolve(pluginDirArg);
  const validator = join(HERE, "validate_goose_plugin.js");
  execFileSync("node", [validator, root], { stdio: "inherit" });

  const output = outputArg
    ? resolve(outputArg)
    : join(dirname(root), `${basename(root)}.zip`);
  mkdirSync(dirname(output), { recursive: true });

  const files: string[] = [];
  collectFiles(root, root, files);

  const zip = new AdmZip();
  const rootName = basename(root);
  for (const path of files) {
    if (shouldExclude(path, root) || resolve(path) === output) continue;
    const relative = path.slice(root.length + 1);
    const arcname = join(rootName, relative);
    zip.addLocalFile(path, dirname(arcname));
  }
  zip.writeZip(output);
  console.log(output);
}

main();
